pkg/policy/api/rule.go

```diff
@@ -0,0 +1,234 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"github.com/cilium/cilium/pkg/labels"
+)
+
+// Rule is a policy rule which must be applied to all endpoints which match the
+// labels contained in the endpointSelector
+//
+// Multiple rules can select the same endpoint in which case all rules are
+// accumulated additively
+//
+// The rule is split into an ingress section which contains all rules
+// applicable at ingress, and an egress section applicable at egress.
+//
+// Either ingres, egress, or both can be provided. If both ingress and egress
+// are omitted, the rule has no effect.
+type Rule struct {
+	// EndpointSelector selects all endpoints which should be subject to
+	// this rule. Cannot be empty.
+	EndpointSelector EndpointSelector `json:"selector"`
+
+	// Ingress contains all rule types which can be applied at ingress
+	//
+	// If omitted, no rules will be applied at ingress via this rule.
+	//
+	// +optional
+	Ingress *IngressRule `json:"ingress,omitempty"`
+
+	// Egress contains all rule types which can be applied at egress
+	//
+	// If omitted, no rules will be applied at egress via this rule
+	//
+	// +optional
+	Egress *EgressRule `json:"egress,omitempty"`
+
+	// Labels is a list of optional strings which can be used to
+	// re-identify the rule or to store metadata. It is possible to lookup
+	// or delete strings based on labels. Labels are not required to be
+	// unique, multiple rules can have overlapping or identical labels.
+	//
+	// +optional
+	Labels labels.LabelArray `json:"labels,omitempty"`
+
+	// Comment is a free form string, it can be used by the creator of the
+	// rule to store human readable explanation of the purpose of this
+	// rule. Rules cannot be identified by comment.
+	//
+	// +optional
+	Comment string `json:"comment,omitempty"`
+}
+
+// IngressRule contains all rule types which can be applied at ingress,
+// i.e. network traffic that originates outside of the endpoint and
+// is entering the endpoint selected by the endpointSelector.
+//
+// - All members of this structure are optional.
+// - All rule types are considered independently
+type IngressRule struct {
+	// FromEndpoints is a list of endpoints identified by an
+	// EndpointSelector which are allowed to communicate with the endpoint
+	// subject to the rule.
+	//
+	// If omitted and no other IngressRule applies a FromEndpoints rule on
+	// the endpoint, the endpoint is only allowed to take connections from
+	// endpoints which are whitelisted in fromCIDR.
+	//
+	// +optional
+	FromEndpoints []EndpointSelector `json:"fromEndpoints,omitempty"`
```

Comment:
Source has some implications as it could be confused for source addresses. "From" is clearer in that sense, "rule selects b and allows ingress from a"